### PR TITLE
feat: allow serialization/deserialization of custom data types

### PR DESCRIPTION
**Pull request by trueadm**
Enables serialization and deserialization of custom data types in the form of hooks.

Closes https://github.com/sveltejs/kit/issues/9401.

### CODE REVIEW

**Comment by Rich-Harris** [comment]
preview: https://svelte-dev-git-preview-kit-13125-svelte.vercel.app/

_this is an automated message_

**Comment by dominikg** [comment]
I'd love an api where `serialize` and `deserialize` implementations for a custom type are in the same file.

They have a strict relation to one another and changing one without the other can lead to bugs. It would also simplify implementing tests that ensure serialize(deserialize(x)) equals x  and deserialize(serialize(y)) equals y

Users can possibly do that in userland by importing them from such a file, but just lik param matchers i think we could provide that from a known location.

**Comment by trueadm** [comment]
> I'd love an api where `serialize` and `deserialize` implementations for a custom type are in the same file.
> 
> They have a strict relation to one another and changing one without the other can lead to bugs. It would also simplify implementing tests that ensure serialize(deserialize(x)) equals x and deserialize(serialize(y)) equals y
> 
> Users can possibly do that in userland by importing them from such a file, but just lik param matchers i think we could provide that from a known location.

So I tried to do something like that, but I gave up. I couldn't figure out how how we could safely take the function on the server and use it on the client, given the function references imports.

**Comment by dominikg** [comment]
its a pair of pure functions tied together by a name, so shouldn't it work in a similar way as `matchers`?
maybe 3 functions if you want to add a test function too
```ts
export const is(foo: any): boolean;
export const serialize(foo: Foo): string;
export const deserialize(str: string): Foo;
```
due to treeshaking only one of serialize and deserialize should make it to the output of client or server.

**Comment by trueadm** [comment]
> its a pair of pure functions tied together by a name, so shouldn't it work in a similar way as `matchers`? maybe 3 functions if you want to add a test function too
> 
> ```ts
> //src/serializers/foo.js
> export const is(foo: any): boolean;
> export const serialize(foo: Foo): string;
> export const deserialize(str: string): Foo;
> ```
> 
> due to treeshaking only one of serialize and deserialize should make it to the output of client or server.

The body of the functions will require importing the class though. So when it runs on the client, the import will be missing?

File: packages/kit/test/apps/basics/test/test.js [hunk 1]
```diff
@@ -1379,3 +1379,24 @@ test.describe.serial('Cookies API', () => {
 		expect(await span.innerText()).toContain('undefined');
 	});
 });
+
+test.describe('Serialization', () => {
+	test('A custom data type can be serialized/deserialized', async ({ page }) => {
+		await page.goto('/serialization-basic');
+		expect(await page.textContent('h1')).toBe('It works!');
+	});
```
Comment (dummdidumm) [inline packages/kit/test/apps/basics/test/test.js:1387]: We need another test that starts somewhere else and then navigates to the page client-side (e.g. you add a link to `/serialization-basic` from some other page, start on said page and make playwright click on that link) in order to test that `_data.json` is also serialized

File: packages/kit/src/runtime/server/page/render.js [hunk 2]
```diff
@@ -322,11 +322,17 @@ export async function render_response({
...
 
 			properties.push(`resolve: ({ id, data, error }) => {
-							const { fulfil, reject } = deferred.get(id);
-							deferred.delete(id);
-
-							if (error) reject(error);
-							else fulfil(data);
+							const try_to_resolve = () => {
+								if (!deferred.has(id)) {
+									setTimeout(try_to_resolve, 0);
```
Comment (dummdidumm) [inline packages/kit/src/runtime/server/page/render.js:327]: was is necessary after all due to the dynamic imports? If so we should add a comment here stating why this is necessary.

**Review by dummdidumm — COMMENTED** [review]
Looks good so far! I think this doesn't handle the client side navigation case yet though, i.e. when the data loading happens via `__data.json`. E.g. if you start on `/foo` and that contains a link to `/serialization-basic`, so it (de)serialize the data correctly?

**Comment by trueadm** [comment]
@dummdidumm Thanks for the feedback. Fixed!

File: packages/kit/src/runtime/server/page/render.js (hunk 2 above)
Comment (Rich-Harris) [reply packages/kit/src/runtime/server/page/render.js:329]: this feels suuuuuper hacky, surely we can find another spproach?

**Comment by Rich-Harris** [comment]
Opened an alternative PR that puts the encoding/decoding logic in a single hook: #13149

**Comment by dummdidumm** [comment]
Closing in favor of #13149

**Pull request closed without merging**